run/runner.py

```diff
@@ -0,0 +1,387 @@
+import gzip
+import json
+import os
+import requests
+import subprocess
+import time
+
+
+# TODO after --install-browser verify browser
+#      version or change platform ID to correct browser version
+#      This can be done by running `wpt install firefox` and then verifying
+# TODO report OS version when creating TestRun
+# TODO check out correct WPT revision
+# TODO convert these back into command line args. They're currently
+#      env vars due to an earlier change. They should be converted back.
+# TODO break this file up into smaller, unit tested modules.
+# TODO document how we use metadata
+class Runner(object):
+    ASSERT_WPT_PATH = (
+        'Runner.wpt_path required. (Default: `WPT_PATH` env var.)'
+    )
+    ASSERT_WPT_PATH_FMT = 'Runner.wpt_path cannot end with /. %s' % (
+        '(Default: `WPT_PATH` env var.)',
+    )
+    ASSERT_SHA_LEN = 'Runner.sha required. (Default: `WPT_SHA` env var.)'
+
+    def __init__(
+        self,
+
+        # For making instance-internal network requests for metadata.
+        # See:
+        # https://cloud.google.com/compute/docs/storing-retrieving-metadata
+        metadata_url=None,
+        prod_host=None,
+        gs_results_bucket=None,
+        prod_run=False,
+        prod_wet_run=False,
+        sha=None,
+        sauce_from_metadata=False,
+        sauce_key=None,
+        sauce_user=None,
+        sauce_tunnel_id=None,
+        wpt_path=None,
+        wptd_path=None,
+        output_path=None,
+        # Passing RUN_PATH will run a specific path in WPT (e.g. html)
+        run_path=None,
+        upload_secret=None,
+        platform_id=None,
+        platform=None,
+        local_log_filepath=None,
+        local_report_filepath=None,
+        summary_path=None,
+        local_summary_gz_filepath=None,
+        gs_results_filepath_base=None,
+        gs_http_results_url=None,
+        summary_filename=None,
+        summary_http_url=None,
+    ):
+        self.metadata_url = metadata_url
```

Comment:
nit: I think we should have a plan to document what all of these are, maybe just a short sentence but would be nice to have in the future. Most of these are self explanatory by the name, but also requires a little context.

Reply:
I agree, we should have a follow-up PR for that. I'm not convinced this interface is "mature" yet. Some config to control push-to-BigQuery (vs. / alongside push-to-GCS) will be added soon, for example.